llvm/lib/CodeGen/EVLIndVarSimplify.cpp

```diff
@@ -0,0 +1,296 @@
+    if (&PN == IndVar)
+      continue;
+
+    // Check 1: it has to contain both incoming (init) & backedge blocks
+    // from IndVar.
+    if (PN.getBasicBlockIndex(InitBlock) < 0 ||
+        PN.getBasicBlockIndex(BackEdgeBlock) < 0)
+      continue;
+    // Check 2: EVL index is always increasing, thus its inital value has to be
+    // equal to either the initial IV value (when the canonical IV is also
+    // increasing) or the last IV value (when canonical IV is decreasing).
+    Value *Init = PN.getIncomingValueForBlock(InitBlock);
+    using Direction = Loop::LoopBounds::Direction;
+    switch (Bounds->getDirection()) {
+    case Direction::Increasing:
+      if (Init != CanonicalIVInit)
+        continue;
+      break;
+    case Direction::Decreasing:
+      if (Init != CanonicalIVFinal)
+        continue;
+      break;
+    case Direction::Unknown:
+      // To be more permissive and see if either the initial or final IV value
+      // matches PN's init value.
+      if (Init != CanonicalIVInit && Init != CanonicalIVFinal)
+        continue;
+      break;
+    }
+    Value *RecValue = PN.getIncomingValueForBlock(BackEdgeBlock);
+    assert(RecValue);
+
+    LLVM_DEBUG(dbgs() << "Found candidate PN of EVL-based IndVar: " << PN
+                      << "\n");
+
+    // Check 3: Pattern match to find the EVL-based index and total trip count
+    // (TC).
+    if (match(RecValue,
+              m_c_Add(m_ZExtOrSelf(IntrinsicMatch), m_Specific(&PN))) &&
+        match(RemTC, m_Sub(m_Value(TC), m_Specific(&PN)))) {
+      EVLIndVar = RecValue;
+      break;
+    }
+  }
+
+  if (!EVLIndVar || !TC)
+    return false;
+
+  // Make sure TC is related to the original trip count of the canonical IV.
+  // Specifically, if the canonical trip count is derived from TC.
+  const SCEV *TCV = SE.getSCEV(TC);
+  bool MatchTC = false;
+  if (const auto *ConstTCV = dyn_cast<SCEVConstant>(TCV)) {
+    // If TC is a constant and vscale is also a constant, then the canonical
+    // trip count will be constant. Canonical trip count * Step equals to the
+    // round up of TC.
+    if (const auto *ConstStep = dyn_cast<SCEVConstant>(StepV))
+      if (unsigned CanonicalTC = SE.getSmallConstantTripCount(&L)) {
+        APInt Step = ConstStep->getAPInt().abs().zextOrTrunc(64);
+        APInt CanonicalTripCount(64, CanonicalTC);
+        APInt TripCount = ConstTCV->getAPInt().zextOrTrunc(64);
+        MatchTC = (CanonicalTripCount * Step - TripCount).ult(Step);
+      }
+  }
+  // Otherwise, we simply check if the upper or lower bound expression of the
+  // canonical IV contains TC.
+  auto equalsTC = [&](const SCEV *S) -> bool { return S == TCV; };
+  if (!MatchTC && !llvm::SCEVExprContains(CanonicalIVFinalV, equalsTC) &&
+      !llvm::SCEVExprContains(CanonicalIVInitV, equalsTC))
+    return false;
+
+  LLVM_DEBUG(dbgs() << "Using " << *EVLIndVar << " for EVL-based IndVar\n");
+
+  // Create an EVL-based comparison and replace the branch to use it as
```

Comment:
I want to suggest an alternate approach here.  It took me quite a while to try to understand what you're trying to prove here, and I think there's an easier and more general alternative.

First, terminology.  EVLStep = min(VF, sub_nsw(EVL_TC, IV))

Step 1 - Check that the step of the IV being replaced is the same as the EVL IV's step on all but the last iteration.  (You already do this, thought the comments don't explain the significance.)

Step 2 - Using SCEV's backedge taken count reasoning, compute the exit value for the EVL IV.

SCEV won't directly do this for you - that's the whole countability bit - but you should be able to compute it as:
IV->evaluateAtIteration(BTC) + EVLStep.evaluateAtIteration(BTC).  (You could simplify the EVLStep based on the final iteration assumption, but you don't need to, and avoiding it reduces your scope for bugs.)

Note that the fact we know EVLStep <= IV.Step is critical for the correctness of the above.  If we didn't, we'd have to prove non-trivial wrap properties.  

Step 3 - Rewrite the exit condition to be an equality comparison between the EVLIV and the new exit value.  Note that you don't need the equality test restriction from the current code.  An important point is that the exit value above is very likely to be EVL_TC in practice, but that proving that is a proof burden you don't actually need.

Use SCEVExpander for this, and let it worry about all the simplifications.

Reply:
Thank you for the suggestion, I think `IV->evaluateAtIteration(BTC) + EVLStep.evaluateAtIteration(BTC)` is pretty neat.

I have one question here though:
> First, terminology. EVLStep = min(VF, sub_nsw(EVL_TC, IV))

Is it supposed to be EVLStep = min(_IV.Step_, sub_nsw(EVL_TC, IV))?

Reply:
No, it wasn't.  The active_vector_length has the VF parameter.  That *may* be the same as the step, but in principle doesn't have to be.  Note that my (1) above (which your code already does) is proving VF=IV.Step, but that is a step you need to prove.  It's not something you can assume.

Reply:
I think there is another problem here: the backedge taken count cannot be computed when vscale is not constant. Namely, IV.Step is in the most commonly seen form of `(VF x vscale)`. Which also causes a problem in another comment thread where you suggested avoid using `getBounds`, because I guess you were suggesting to get the final IV SCEV via something like `IV.evaluateAtIteration(CanonicalTripCount)`, but CanonicalTripCount, just like backedge taken count, cannot be computed in this (i.e. non-constant vscale) case.

However, alternatively I think we can get exit value via this:
```
IV_BTC = IV_Final - IV.Step
Tail = TC - IV_BTC
Exit value = IV_BTC + Tail
```
`IV_BTC` as the name suggested, is the equivalence of `IV->evaluateAtIteration(BTC)` you suggested, but this time we calculate it from `IV_Final`, which is, well, final IV value returned from `getBounds`. Since it only uses IR pattern matching we don't have to worry about the aforementioned uncomputable problem.

Reply:
> think there is another problem here: the backedge taken count cannot be computed when vscale is not constant. Namely, IV.Step is in the most commonly seen form of `(VF x vscale)`....  just like backedge taken count, cannot be computed in this (i.e. non-constant vscale) case.

Ok, this seems weird.  I don't know of any fundamental reason that SCEV shouldn't be able to figure out a trip count for a loop invariant step - even if non constant.  Would you mind sharing a piece of IR which demonstrates this?  I want to see if this is something easy to fix, or I'm forgetting some subtle interaction.

Reply:
> Would you mind sharing a piece of IR which demonstrates this? I want to see if this is something easy to fix, or I'm forgetting some subtle interaction.

I was using the first function (i.e. `@simple`) in the test for this PR (`evl-iv-simplify.ll`). I actually traced down to why it could not compute: the smoking gun is `ScalarEvolution::howFarToZero`, specifically, [this line](https://github.com/llvm/llvm-project/blob/7dc84e225e11e37925db6f4f08269f447d2f2347/llvm/lib/Analysis/ScalarEvolution.cpp#L10494) that actually bails out with SCEVCouldNotCompute. As the TODO comment right above it suggested, it couldn't handle non-constant step in AddRec for now.

Reply:
> > Would you mind sharing a piece of IR which demonstrates this? I want to see if this is something easy to fix, or I'm forgetting some subtle interaction.
> 
> I was using the first function (i.e. `@simple`) in the test for this PR (`evl-iv-simplify.ll`). I actually traced down to why it could not compute: the smoking gun is `ScalarEvolution::howFarToZero`, specifically, [this line](https://github.com/llvm/llvm-project/blob/7dc84e225e11e37925db6f4f08269f447d2f2347/llvm/lib/Analysis/ScalarEvolution.cpp#L10494) that actually bails out with SCEVCouldNotCompute. As the TODO comment right above it suggested, it couldn't handle non-constant step in AddRec for now.

Thanks!

That limitation actually looks pretty easy to at least partially remove.  I'm currently a bit buried, but expect a patch in the next few days.

Reply:
See https://github.com/llvm/llvm-project/pull/94411

Though this is with two changes to your test:
* adding vscale_range to the function
* making the IV nuw

I think both should apply to your original (i.e. your test is slightly over-reduced), but please confirm.

Reply:
> See #94411
> 
> Though this is with two changes to your test:
> 
> * adding vscale_range to the function
> * making the IV nuw
> 
> I think both should apply to your original (i.e. your test is slightly over-reduced), but please confirm.

Thank you for the quick fix.
I gave it a try and found 1 minor and 1 major concern: the minor one is that `vscale_range` has to be set otherwise BTC cannot be computed. This is because when we're computing BTC,  one of the preconditions is that the range of `(VF x vscale)` has to be non-zero, which depends on the range of vscale. Currently we use `[1,0)` ,a.k.a wrap around starting from 1, to represent the range of vscale when `vscale_range` is not set -- but this is an _unsigned_ wrap around. However, when computing the multiplication range of `(VF x vscale)` this `[1,0)` is treated as _signed_ wrap around, leading to `(VF x vscale)`'s range being full_set. Therefore when being asked the minimum unsigned value of `(VF x vscale)`'s range, it returns zero. I call this "minor" because I believe in practice vscale_range will mostly be set. But it's still a little annoyed to be that we cannot represent explicit unsigned range like `[1, UINT_MAX]` with `ConstantRange`.

The major concern being that the instructions for computing exit value expanded from the SCEV expressions we crafted is pretty verbose, long, a more importantly, contain division instructions. Let me start with the formula we have:
```
IV = the original canonical IV
EVL_TC = the trip count in the scalar loop
IV.Step = (VF x vscale)
EVLStep = umin(IV.Step, sub_nsw(EVL_TC, IV))
BTC = backedge taken count in the original canonical IV loop
ExitValue = IV.evaluateAtIteration(BTC) + EVLStep.evaluateAtIteration(BTC)
```
In your original formula EVLStep is `min(VF, sub_nsw(EVL_TC, IV))`, which I had a question on whether the first operand of `min` is VF or `IV.Step`. Because to my best understandings, `VF` is the same VF we put in the scalable vector type we use, `(vscale x VF x ...)`, and also the VF used in computing step value in IV: `(VF x vscale)`. So I think we should use IV.Step here. For the very least, if I use `VF` (as in my definition) here, the result will be pretty wrong.

Now, since only SCEVAddRec has `evaluateAtIteration`, we can't really do `EVLStep.evaluateAtIteration`. Instead, I rewrote `EVLStep.evaluateAtIteration` into:
```
umin(IV.Step, sub_nsw(EVL_TC, IV.evaluateAtIteration(BTC))
```
Because IV.Step and EVL_TC are both loop invariant.
Using the `@simple` function in my test case + the same modifications you made in your comment above, as an example, this final `IV.evaluateAtIteration(BTC) + umin(IV.Step, sub_nsw(EVL_TC, IV.evaluateAtIteration(BTC))` looks like:
```
((4 * vscale * ((-1 * vscale * (4 + (-4 * ((-1 + (4 * vscale)<nuw><nsw> + %N) /u (4 * vscale)<nuw><nsw>))<nsw>)<nsw>) /u (4 * vscale)<nuw><nsw>)) + (((-4 * vscale * ((-1 * vscale * (4 + (-4 * ((-1 + (4 * vscale)<nuw><nsw> + %N) /u (4 * vscale)<nuw><nsw>))<nsw>)<nsw>) /u (4 * vscale)<nuw><nsw>)) + %N) umin (4 * vscale)<nuw><nsw>))
```
And those divisions seen in the expressions will be expanded into division instructions. I'm concern whether we should spend so many instructions to compute something that will almost certain be `EVL_TC`.

I also run the same algorithm on the `@fixed_iv_step_tc` function in my test case, which has fixed IV.Step and fixed trip count. The generated exit value is correct, which means my methodology is probably not terribly off.

I've also tried to use `sub_nsw(EVL_TC, IV.evaluateAtIteration(BTC)`, which is the tail in the last iteration, in replacement of `umin(IV.Step, sub_nsw(EVL_TC, IV.evaluateAtIteration(BTC))`. But the resulting Exit value, `IV.evaluateAtIteration(BTC) + sub_nsw(EVL_TC, IV.evaluateAtIteration(BTC)` is not really meaningful (it will certainly be `EVL_TC` though).

Last but not the least, I'm still a little confused about how to do Step (1) you described: it's true that I already did such a check -- but only on cases where both trip count and vscale are constant. For every other cases we only have SCEV expressions, which we cannot know it value during compile time. Even we expand the check into runtime check, what should we do if the check fails during runtime? Do we fall back the original canonical IV?

Reply:
> > See #94411
> > Though this is with two changes to your test:
> > 
> > * adding vscale_range to the function
> > * making the IV nuw
> > 
> > I think both should apply to your original (i.e. your test is slightly over-reduced), but please confirm.
> 
> Thank you for the quick fix. I gave it a try and found 1 minor and 1 major concern: the minor one is that `vscale_range` has to be set otherwise BTC cannot be computed. This is because when we're computing BTC, one of the preconditions is that the range of `(VF x vscale)` has to be non-zero, which depends on the range of vscale. Currently we use `[1,0)` ,a.k.a wrap around starting from 1, to represent the range of vscale when `vscale_range` is not set -- but this is an _unsigned_ wrap around. However, when computing the multiplication range of `(VF x vscale)` this `[1,0)` is treated as _signed_ wrap around, leading to `(VF x vscale)`'s range being full_set. Therefore when being asked the minimum unsigned value of `(VF x vscale)`'s range, it returns zero. I call this "minor" because I believe in practice vscale_range will mostly be set. But it's still a little annoyed to be that we cannot represent explicit unsigned range like `[1, UINT_MAX]` with `ConstantRange`.

Clang emits the vscale_range by default.  vscale is bounded by the values of possible VLEN, and VLEN has a specific upper bound in the RISCV vector specification.  So while annoying that SCEV can't just ask the backend for this, this is a non-problem for any real c/c++ input.  

For the rest of your response, one quick comment - the divisions are by 4 * vscale.  We know that vscale is a power of two by definition, so these are right shifts.  You do need a popcount to know which power of two, but we could reasonable emit one popcount and a bunch of shifts.  (Not saying we do so today - I'm saying it's possible.)

For the rest, I feel we are talking past each other - and frankly, I'm having trouble following the detail in github PR interface - can I suggest we setup a brief phone call to chat this through instead?

Reply:
> For the rest, I feel we are talking past each other - and frankly, I'm having trouble following the detail in github PR interface - can I suggest we setup a brief phone call to chat this through instead?

That will be really helpful! I'll send you an email to your "public" email address to coordinate this.